File: src/main/java/com/timepath/hl2/hudeditor/PropertyTable.java
package com.timepath.hl2.hudeditor;

import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import javax.swing.table.DefaultTableCellRenderer;
import javax.swing.table.DefaultTableModel;
import javax.swing.table.TableCellRenderer;
import java.awt.*;

/**
 * @author TimePath
 */
@SuppressWarnings("serial")
public class PropertyTable extends JTable {

    @NotNull
    private final CustomTableCellRenderer renderer;
    @NotNull
    private final DefaultTableModel model;

    public PropertyTable() {
        this.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
        model = new DefaultTableModel();
        model.addColumn(Main.getString("Key"));
        model.addColumn(Main.getString("Value"));
        model.addColumn(Main.getString("Info"));
        this.setModel(model);
        renderer = new CustomTableCellRenderer();
    }

    @NotNull
    public DefaultTableModel getModel() {
        return model;
    }

    public void clear() {
        for (int i = model.getRowCount() - 1; i >= 0; i--) {
            model.removeRow(i);
        }
        scrollRectToVisible(new Rectangle(0, 0, 0, 0));
    }

    /**
     * Prevent editing of key
     */
    @Override
    public boolean isCellEditable(int row, int column) {
        return column != 0;
    }

    @NotNull
    @Override
    public TableCellRenderer getCellRenderer(int row, int column) {
        return renderer;
    }

    private class CustomTableCellRenderer extends DefaultTableCellRenderer {

        @NotNull
        @Override
        public Component getTableCellRendererComponent(JTable table,
                                                       Object value,
                                                       boolean isSelected,
                                                       boolean hasFocus,
                                                       int row,
                                                       int column) {
            super.getTableCellRendererComponent(table, value, isSelected, hasFocus, row, column);
            if (value instanceof ImageIcon) {
                @NotNull Icon icon = (Icon) value;
                setText("");
                setIcon(icon);
                setRowHeight(row, Math.max(icon.getIconHeight(), rowHeight));
            } else {
                setText((value == null) ? "" : value.toString());
                setIcon(null);
            }
            return this;
        }
    }
}

File: src/main/kotlin/com/timepath/hl2/hudeditor/PropertyTable.kt
package com.timepath.hl2.hudeditor


import java.awt.Component
import java.awt.Rectangle
import javax.swing.ImageIcon
import javax.swing.JTable
import javax.swing.ListSelectionModel
import javax.swing.table.DefaultTableCellRenderer
import javax.swing.table.DefaultTableModel
import javax.swing.table.TableCellRenderer

public class PropertyTable : JTable() {

    private val renderer: CustomTableCellRenderer
    private val model: DefaultTableModel

    init {
        this.setSelectionMode(ListSelectionModel.SINGLE_SELECTION)
        model = DefaultTableModel()
        model.addColumn(Main.getString("Key"))
        model.addColumn(Main.getString("Value"))
        model.addColumn(Main.getString("Info"))
        this.setModel(model)
        renderer = CustomTableCellRenderer()
    }

    override fun getModel(): DefaultTableModel {
        return model
    }

    public fun clear() {
        run {
            var i = model.getRowCount() - 1
            while (i >= 0) {
                model.removeRow(i)
                i--
            }
        }
        scrollRectToVisible(Rectangle(0, 0, 0, 0))
    }

    /**
     * Prevent editing of key
     */
    override fun isCellEditable(row: Int, column: Int): Boolean {
        return column != 0
    }

    override fun getCellRenderer(row: Int, column: Int): TableCellRenderer {
        return renderer
    }

    private inner class CustomTableCellRenderer : DefaultTableCellRenderer() {

        override fun getTableCellRendererComponent(table: JTable, value: Any?, isSelected: Boolean, hasFocus: Boolean, row: Int, column: Int): Component {
            super.getTableCellRendererComponent(table, value, isSelected, hasFocus, row, column)
            if (value is ImageIcon) {
                setText("")
                setIcon(value)
                setRowHeight(row, Math.max(value.getIconHeight(), rowHeight))
            } else {
                setText(when (value) {
                    null -> ""
                    else -> value.toString()
                })
                setIcon(null)
            }
            return this
        }
    }
}
